feat: Docs for client secret rotation (#2424)
ory/docs: docs/hydra/guides/oauth2-clients.mdx

Reviewer: :::info

Clients that use the `client_secret_basic` or `client_secret_post` `token-endpoint-auth-method` are able to perform secret rotation.

:::

Reviewer: :::warning

If you use secret rotation, setting a new secret through the patch client APIs clears all rotated secrets.

:::

Reviewer: :::warning

If you use secret rotation, setting a new secret through the update client APIs clears all rotated secrets.

:::

</Tabs>
````

## Rotate OAuth2 client secret

OAuth2 client secret rotation allows you to change a client's secret without downtime. When you rotate a secret, the old secret
remains valid until you explicitly clean it up, allowing you to update all services using the client credentials without service
interruption.

### How secret rotation works

1. **Rotate the secret**: Generate a new secret for the client
2. **Both secrets work**: Old and new secrets both authenticate until cleanup
3. **Update services**: Update your applications to use the new secret

> 4. **Cleanup**: Manually remove old rotated secrets once all services are updated

Reviewer 2: ```suggestion
1. Rotate the secret: Generate a new secret for the client
2. Both secrets work: Old and new secrets both authenticate until cleanup
3. Update services: Update your applications to use the new secret
4. Cleanup: Manually remove old rotated secrets once all services are updated
```

</Tabs>
````

## Rotate OAuth2 client secret

OAuth2 client secret rotation allows you to change a client's secret without downtime. When you rotate a secret, the old secret
remains valid until you explicitly clean it up, allowing you to update all services using the client credentials without service
interruption.

##### How secret rotation works

> 1. Rotate the secret: Generate a new secret for the client

Reviewer 3: Done — the step now reads “Generate a new secret for the client service”.

> 2. Both secrets work: Old and new secrets both authenticate until cleanup

Reviewer 3: Addressed with your newer suggestion: the step now reads “Both the old and the new secrets can be used to authenticate from this point on”, which keeps the fact that dual validity starts at rotation while making clear services still need the new secret before they can use it.

> 3. Update services: Update your applications to use the new secret

Reviewer 3: Done — “Update your client services to use the new secret.”

> 4. Cleanup: Manually remove old rotated secrets once all services are updated

Reviewer: ```suggestion
4. After you test the client services can authenticate with the new secret, manually remove the old secret.
```

Reviewer 3: Done — split into “Test that the client services can authenticate with the new secret” and “Manually remove the old secret”.

> ##### Rotate client secret

Reviewer 3: Done — heading changed to “Rotate OAuth2 client secret”.

To rotate an OAuth2 client secret, use the following methods:

````mdx-code-block
<Tabs>
<TabItem value="rest" label="REST API">

```bash
curl -X POST https://{project.slug}.projects.oryapis.com/admin/clients/{client-id}/secret/rotate \
  -H "Authorization: Bearer ory_pat_..."
```

> The response includes the new `client_secret`. Save this value immediately - it will not be shown again.

Reviewer 3: The docs now say: “Store it immediately in the secret store that your client services read from, such as a secrets manager or an encrypted environment configuration.” Ory only shows the secret once, so where exactly it goes depends on the user's setup — we point them at a secrets manager as the recommended option.

See [API documentation](../../reference/api#tag/oAuth2/operation/rotateOAuth2ClientSecret).

</TabItem>
<TabItem value="sdk" label="Ory SDK">

```typescript
import { Configuration, OAuth2Api } from "@ory/client"

const ory = new OAuth2Api(
  new Configuration({
    basePath: `https://${projectSlug}.projects.oryapis.com`,
    accessToken: "ory_pat_..."
  })
)

const { data: client } = await ory.rotateOAuth2ClientSecret({
  id: clientId
})

// Save the new client_secret immediately
console.log("New secret:", client.client_secret)
```

</TabItem>
</Tabs>
````

> ##### Clear rotated secrets

Reviewer 3: Done — heading is now “Remove old secret”.

> Once all services have been updated to use the new secret, you can remove the old rotated secrets to revoke access using the old

Reviewer 3: Done — applied.

> credentials:

Reviewer 3: Done — applied.

````mdx-code-block
<Tabs>
<TabItem value="rest" label="REST API">

```bash
curl -X DELETE https://{project.slug}.projects.oryapis.com/admin/clients/{client-id}/secret/rotate \
  -H "Authorization: Bearer ory_pat_..."
```

> After cleanup, only the current secret will be valid. Old secrets will no longer authenticate.

Reviewer 3: Done — applied.

See [API documentation](../../reference/api#tag/oAuth2/operation/deleteRotatedOAuth2ClientSecrets).

</TabItem>
<TabItem value="sdk" label="Ory SDK">

```typescript
await ory.deleteRotatedOAuth2ClientSecrets({
  id: clientId
})

> // Old secrets are now revoked

Reviewer 3: Done — applied.

```

</TabItem>
</Tabs>
````

##### Secret rotation workflow example

Here's a complete workflow for rotating a client secret:

```bash
# 1. Get current client
CLIENT_ID="your-client-id"

# 2. Rotate the secret
NEW_SECRET=$(curl -X POST "https://{project.slug}.projects.oryapis.com/admin/clients/$CLIENT_ID/secret/rotate" \
  -H "Authorization: Bearer ory_pat_..." | jq -r '.client_secret')

echo "New secret: $NEW_SECRET"

> # 3. Update your applications with the new secret

Reviewer 3: Done — applied (with “services” spelled out).

# (Both old and new secrets work during this period)

# 4. Verify the new secret works
curl -X POST "https://{project.slug}.projects.oryapis.com/oauth2/token" \
  -u "$CLIENT_ID:$NEW_SECRET" \
  -d "grant_type=client_credentials"

> # 5. Once all services are updated, clean up old secrets

Reviewer 3: Done — applied.

curl -X DELETE "https://{project.slug}.projects.oryapis.com/admin/clients/$CLIENT_ID/secret/rotate" \
  -H "Authorization: Bearer ory_pat_..."

# Old secret is now revoked
```

:::tip Zero-downtime credential rotation

> Secret rotation enables zero-downtime credential updates. Both the old and new secrets remain valid until you manually clean up

Reviewer: ```suggestion
Secret rotation enables zero-downtime credential updates. Both the old and new secrets remain valid until you manually remove
```

Reviewer 3: Done — applied.

> the rotated secrets, allowing you to update all your services without service interruption.

Reviewer 3: Done — applied.

:::

:::warning Security best practice

> Rotated secrets remain valid indefinitely until you explicitly clean them up. Always remove old rotated secrets once your

Reviewer 3: Done — applied.

> Rotated secrets remain valid indefinitely until you explicitly clean them up. Always remove old rotated secrets once your

Reviewer 3: Confirmed against the implementation (ory-corp/cloud#10632): OAuth2 client secrets have no expiry or validity date. Rotated secrets stay valid until you delete them via the DELETE endpoint, or until they age out — only the five most recent rotated secrets are kept per client (now documented in an info box). The `valid_to` attribute you mention is for SAML signing certificates and doesn't apply to OAuth2 client secrets.

> migration is complete to ensure that compromised credentials cannot be used.

Reviewer 3: Done — applied.

## Rotate OAuth2 client secret

OAuth2 client secret rotation allows you to change a client's secret without downtime. When you rotate a secret, the old secret
remains valid until you remove it, allowing you to update all your client services without service interruption.

##### How secret rotation works

> 1. Generate a new secret for the client service. Both the old and the new secret authenticate from this point on.

Reviewer: Secret needs to be plural: Both the old and the new secrets can be used to authenticate from this point on

Reviewer: I think AI was used to reply/update the original comment?: "That would imply the dual-validity window starts once you update your services. It actually starts the moment you rotate."

But it missed the point that authentication with the new secret cannot happen if the client is not updated with the new secret. Both old and new secrets exist and can be used, but only after the update service step will it actually authenticate anything.

"Both the old and the new secrets **can be used** to authenticate from this point on."

Reviewer 3: Applied your exact wording: “Both the old and the new secrets can be used to authenticate from this point on.” Thanks — you're right that authentication with the new secret only happens once services are updated; “can be used” captures that.

</Tabs>
````

## Rotate OAuth2 client secret

OAuth2 client secret rotation allows you to change a client's secret without downtime. When you rotate a secret, the old secret
remains valid until you remove it, allowing you to update all your client services without service interruption.

> 

Reviewer: ```suggestion
Secret rotation is available only for clients that use the `client_secret_basic` or `client_secret_post` token endpoint authentication method. Ory keeps up to five rotated secrets per client. When you rotate again, the oldest rotated secret is dropped and can no longer authenticate.

:::info

Setting a new secret through the update or patch client APIs clears all rotated secrets.

:::

```

##### How secret rotation works

> 1. Generate a new secret for the client service. Both the old and the new secrets can be used to authenticate from this point on.

Reviewer: ```suggestion
1. Rotation generates a new secret for the client service. Both the old and the new secrets can be used to authenticate from this point on.
```

2. Update your client services to use the new secret.
3. Test that the client services can authenticate with the new secret.

> 4. Manually remove the old secret. Only the new secret authenticates from this point on.

Reviewer: ```suggestion
4. Delete the old secret. Only the new secret authenticates from this point on.
```

> :::info

Reviewer: ```suggestion

```

> 

Reviewer: ```suggestion
```

> Secret rotation is available only for clients that use the `client_secret_basic` or `client_secret_post` token endpoint

Reviewer: ```suggestion
```

> authentication method. Ory keeps up to five rotated secrets per client. When you rotate again, the oldest rotated secret is

Reviewer: ```suggestion
```

> dropped and can no longer authenticate. Setting a new secret through the update or patch client APIs clears all rotated secrets.

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> :::

Reviewer: ```suggestion
```

</TabItem>
<TabItem value="sdk" label="Ory SDK">

```typescript
await ory.deleteRotatedOAuth2ClientSecrets({
  id: clientId
})

// Old secret is now revoked.
```

</TabItem>
</Tabs>
````

##### Secret rotation workflow example

Here's a complete workflow for rotating a client secret:

```bash
# 1. Get current client
CLIENT_ID="your-client-id"

# 2. Rotate the secret
NEW_SECRET=$(curl -X POST "https://{project.slug}.projects.oryapis.com/admin/clients/$CLIENT_ID/secrets/rotate" \
  -H "Authorization: Bearer ory_pat_..." | jq -r '.client_secret')

echo "New secret: $NEW_SECRET"

# 3. Update your client services with the new secret
# (Both the old and the new secrets work during this period)

# 4. Verify the new secret works
curl -X POST "https://{project.slug}.projects.oryapis.com/oauth2/token" \
  -u "$CLIENT_ID:$NEW_SECRET" \
  -d "grant_type=client_credentials"

# 5. Once all client services are updated, remove the old secret
curl -X DELETE "https://{project.slug}.projects.oryapis.com/admin/clients/$CLIENT_ID/secrets/rotate" \
  -H "Authorization: Bearer ory_pat_..."

# Old secret is now revoked
```

:::tip Zero-downtime credential rotation

> Secret rotation enables zero-downtime credential updates. Both the old and new secrets remain valid until you manually remove the

Reviewer: ```suggestion
Secret rotation enables zero-downtime credential updates. Both the old and new secrets remain valid until you remove the
```

CLIENT_ID="your-client-id"

# 2. Rotate the secret
NEW_SECRET=$(curl -X POST "https://{project.slug}.projects.oryapis.com/admin/clients/$CLIENT_ID/secrets/rotate" \
  -H "Authorization: Bearer ory_pat_..." | jq -r '.client_secret')

echo "New secret: $NEW_SECRET"

# 3. Update your client services with the new secret

> # (Both the old and the new secret work during this period)

Reviewer: Secret should be plural: secrets

Reviewer 3: Done — pluralized to “Both the old and the new secrets work during this period”.
